EngiMarket Part Four - Ranged RCD and Rapid Lighting Device (RLD)

Robustin opened the pull request:
Video worth a thousand words: https://youtu.be/ewMSUMExjVY

Two new items, one brand new, the other an RCD that works at range without any other bells and whistles. 

The RLD allows you to create floor and wall lighting along with temporary (they use fuel like flares) glowsticks at range. It can also deconstruct lights. The best part of this feature (besides the visual "ammo" updating icons) is the ability to select absolutely any color you desire. 

:cl: Robustin
add: Ranged RCD added to the game
add: Rapid Lighting Device (also ranged) has been added. It can create wall, floor, and temporary lights of any color you desire. 
/:cl:

Robustin committed:
Fixing the master

Robustin committed:
Update tgstation.dme

Robustin committed:
ARCD and RLD

Robustin committed:
The joys of atomization

GunHog commented:
:+1: because they look like INDUSTRIAL SMGs!!

MMMiracles commented:
Yeah this is pretty damn cool.

Robustin committed:
Atomization is hell

Robustin committed:
Hell is actually an understatement

Armhulen commented:
>RANGED RCD IS JUST SHIT COLORED

PKPenguin321 commented:
error noise at the very end like last 2 seconds IS WAY TOO FUCKING LOUD MY EARS

Robustin commented:
Yea someone get me a ranged RCD icon plz thx.

Ill turn the error noise down.

Robustin committed:
Save PKP's hearing

FantasticFwoosh commented:
🆑 Robustin
add: This long range RCD was held onto by your daddy in the warcamp, the way he saw it he wasn't going to let those crooks get his hands on it; your birthright, he'll be damned if the greasy crooks put his hands on it. So he hid it, one place he could hide something; his ass. for five long years he wore this RCD, up his ass. Now it belongs to you after he died of dysentry.
/🆑

Dorsisdwarf commented:
Maybe you could use those advanced RCD sprites someone made a while back for the ranged rcd rather than a pink recolor

Robustin commented:
I can't find any reference to an an advanced RCD in the PR's

AdamElTablawy commented:
Mech's RCD is ranged, isn't it? Feels like you're taking away some unique powers from an already rare mech weapon. Granted, this isn't in game but we all know it'll be added in at some point.

GunHog commented:
The mech RCD is indeed ranged, and should be updated to the current specs so it can do things like configure access, customize modes, etc.

Dorsisdwarf commented:
Its not in the game, its in one of the sprite threads of the forum, I'll go nab it

Iamgoofball commented:
mecha tards pls go

Iamgoofball commented:
:+1: good PR

kevinz000 commented:
RLD is fine but ranged RCD only will lead to extraordinary amounts of grief and bullshit.

Robustin commented:
I disagree Kevinz, the ranged RCD has a beam that makes it very obvious what youre up to. If someone if RCD'ing an airlock then then there's only one possible thing they could be doing do it. 

The difference is a normal RCD doesn't telegraph much at all, someone stands near a door, has a little sprite in their hand, and 2 seconds later your airlocks are gone/floor is gone/escape is walled/etc.

The ranged RCD puts a giant target on whoever is doing shit and the tradeoff for convenience is putting a giant red arrow over your head as the person responsible. They're good RCD's but I've found that they are far more useful for construction than destruction.

GunHog commented:
Would you be up to updating the mech RCD with the customization options, Robustin? Probably not in this PR though

kevinz000 commented:
@Robustin how about this people can alt click or something​ on the tile to disrupt progress? It can be used to repair but it will be a bit harder to blue space grief.

Robustin commented:
Im clueless about mechs and I'm still busy on my final feature and rewards PR, after that I hope to get a good deal of testing in and if the ranged RCD turns into a weapon of mass grief that people can't deal with... I'll come up with a solution then.

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+
+
+/obj/item/weapon/rcd/arcd/afterattack(atom/A, mob/user, proximity)
+	proximity = 1
```
Override this with some parameter in base rcd instead.

Robustin replied:
couldn't I just put this in the args? I'm not exactly sure what youre suggesting, all the "camera RCD's" use the proximity override too so I need to be 100% sure before I go changing it.

GunHog replied:
In base RCD: add something like `var/ranged = FALSE` and set it to true for your ranged RCDs, then stick it in the proximity check.

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+	if(!(A in view(7, get_turf(user))))
+		to_chat(user, "<span class='warning'>The \'Out of Range\' light on the RCD blinks red.</span>")
+		return 0
+	if(isturf(A) || istype(A, /obj/machinery/door/airlock) || istype(A, /obj/structure/grille) || istype(A, /obj/structure/window) || istype(A, /obj/structure/girder))
```
Base afterattack already returns true on valid targets, no need to do this typecheck.

Robustin replied:
This is to control the beam, not the actual RCD. My extensive local testing confirmed that being able to spam beams on invalid targets was a bad idea.

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+		user.Beam(A,icon_state="rped_upgrade",time=20)
+		..()
+	else
+		return
```
No need for these two lines

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+	desc = "A device used to rapidly provide lighting sources to an area."
+	icon = 'icons/obj/tools.dmi'
+	icon_state = "rld-5"
+	opacity = FALSE
```
Opacity,density and anchored are all default to item, no need to set these.

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+		..()
+
+/obj/item/weapon/rld/proc/choose_color()
+	set name = "Choose Light Color"
```
No need for these  set's if you don't use the proc as a verb

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+	..()
+
+	desc = "An RLD. It currently holds [matter]/[max_matter] matter-units."
+	src.spark_system = new /datum/effect_system/spark_spread
```
Remove redundant src. please.

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+	spark_system = null
+	. = ..()
+
+/obj/item/weapon/rld/attackby(obj/item/weapon/W, mob/user, params)
```
This whole block smells like copypaste from RCD, if there is not that much difference making it a subtype would probably be saner.

Robustin replied:
It is but it also uses different materials and changes icons based on how much matter is left so I've found that using a child that has to both add and remove the parent's features tends to leave me with buggy workable solutions.

AnturK replied:
Cmon, these are minimal differences. Add possible ammo list and override update_icon, if the update_icon does not fully update state in base rcd that's just bad code.

Jalleo replied:
It is oldcode

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+	//Change the mode
+	playsound(src.loc, 'sound/effects/pop.ogg', 50, 0)
+	switch(mode)
+		if(1)
```
Defines for modes please.

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+		to_chat(user, "<span class='notice'>You insert [amount_to_use] [S.name] sheets into the RLD. </span>")
+		icon_state = "rld-[round(matter/35)]"
+		update_icon()
+		return 1
```
TRUE

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+		update_icon()
+		return 1
+	to_chat(user, "<span class='warning'>You can't insert any more [S.name] sheets into the RLD!")
+	return 0
```
FALSE

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+			mode = 1
+			to_chat(user, "<span class='notice'>You change RLD's mode to 'Deconstruct'.</span>")
+	if(prob(20))
+		src.spark_system.start()
```
Remove redundant src. please.

AnturK reviewed (changes requested):
I'll review the rest later

Iamgoofball commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +389,285 @@ RCD
+		to_chat(user, "<span class='warning'>The \'Out of Range\' light on the RLD blinks red.</span>")
+		return 0
+	switch(mode)
+		if(1)
```
# >refactor RCDs to not be shit
# >it isnt used at all

make this a subtype of RCDs

Robustin replied:
I haven't looked at your refactor yet, I plan to, but I kinda suck when it comes to making subtypes that actually work as intended

Robustin replied:
Goof you're gonna have to explain how I can make this an effective subtype of RCD's when they do very different things. Otherwise im just going to do what Antur suggested.

Robustin committed:
Update RCD.dm

Robustin committed:
RCD review changes

GunHog commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -398,27 +401,26 @@ RCD
 		to_chat(user, "<span class='warning'>The \'Out of Range\' light on the RCD blinks red.</span>")
 		return 0
-	if(isturf(A) || istype(A, /obj/machinery/door/airlock) || istype(A, /obj/structure/grille) || istype(A, /obj/structure/window) || istype(A, /obj/structure/girder))
+	if((isturf(A) && mode==3) || (isturf(A) && mode==1 && !A.density) || (istype(A, /obj/machinery/door/airlock) && mode==3) || istype(A, /obj/structure/grille) || (istype(A, /obj/structure/window) && mode==3) || istype(A, /obj/structure/girder))
```
Use the defiiiiiiiiiiiiiiiiiiiiiiiiiiiiines!!

Robustin replied:
Thats means I have to add defines to the RCD too, the thing that Goof just refactored.

Robustin committed:
conflict fix

Robustin committed:
Forward

Robustin committed:
ugh

Robustin committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into Future7Test

# Conflicts:
#	icons/mob/inhands/items_lefthand.dmi
#	icons/mob/inhands/items_righthand.dmi

Robustin committed:
finalized

Robustin committed:
Merge branch 'Future7' of https://github.com/Robustin/tgstation into Future7Test2

# Conflicts:
#	code/game/objects/items/weapons/RCD.dm
#	icons/mob/inhands/items_lefthand.dmi
#	icons/mob/inhands/items_righthand.dmi

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +388,278 @@ RCD
+/obj/item/weapon/rcd/arcd/afterattack(atom/A, mob/user, proximity)
+	if(!(A in view(7, get_turf(user))))
+		to_chat(user, "<span class='warning'>The \'Out of Range\' light on the RCD blinks red.</span>")
+		return 0
```
FALSE

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +388,278 @@ RCD
+		to_chat(user, "<span class='warning'>The \'Out of Range\' light on the RCD blinks red.</span>")
+		return 0
+	if((isturf(A) && A.density && mode==RCD_DECONSTRUCT) || (isturf(A) && !A.density) || (istype(A, /obj/machinery/door/airlock) && mode==RCD_DECONSTRUCT) || istype(A, /obj/structure/grille) || (istype(A, /obj/structure/window) && mode==RCD_DECONSTRUCT) || istype(A, /obj/structure/girder))
+		user.Beam(A,icon_state="rped_upgrade",time=30)
```
If this giant IF is really needed at least wrap it in some is_valid_target function, but i'm pretty sure there's base rcd function that already does it.

Robustin replied:
This is solely beam control, I tried relying on the just the RCD check but you could run around spamming the ranged beam effect and cluttering the screen.

I tried to wait for the RCD to return TRUE but then the beam wouldn't fire until after the channeling was done, which defeated the point of the beam... to telegraph to other people that someone is about the alter the turf in an area.

Robustin committed:
Update RCD.dm

Jalleo commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -282,7 +288,7 @@ RCD
 
 /obj/item/weapon/rcd/afterattack(atom/A, mob/user, proximity)
-	if(!proximity)
+	if(!proximity && !ranged)
```
Make this a proc and have the ARCD overide the proc and add in the ranged proc. It is much more efficent.

Jalleo replied:
Example if(foo()) 
foo() 
 if! ranged &&! proximity 
          return true

Robustin replied:
I kind of get it:
1) I dont get why this is more efficient than a simple if-statement

2) Im not sure how the ARCD would "override" the proc

Jalleo replied:
Aka you have one in the parent one in the child let me get you a reference one.

Jalleo commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -29,6 +34,7 @@ RCD
 	var/working = 0
 	var/mode = 1
 	var/canRturf = 0
+	var/ranged = FALSE
```
Move this to the child. With the thing I mentioned above.

Jalleo reviewed (commented):
There is a fair bit that can be made into procs

Jalleo commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -382,3 +388,278 @@ RCD
+/obj/item/weapon/rld/attack_self(mob/user)
+	//Change the mode
+	playsound(src.loc, 'sound/effects/pop.ogg', 50, 0)
+	switch(mode)
```
This might have a more efficent method by having it done in another way but I dont know how.

Jalleo commented on code/modules/power/lighting.dm:
```diff
@@ -653,3 +653,14 @@
 		update()
+
+
+/obj/machinery/light/floor
```
Might I ask why this is being done?

Robustin replied:
We don't have floor lighting and I thought it would be nice for more lighting options?

Jalleo replied:
Uh we do have light tiles?

Robustin replied:
Those really weren't what I was going for. I wanted lights built into the floor, not just another silly colored plating.

Robustin committed:
RCD refactor... again

Robustin committed:
Merge branch 'Future7' of https://github.com/Robustin/tgstation into Future7Test3

# Conflicts:
#	code/game/objects/items/weapons/RCD.dm

Robustin committed:
map RCD paths updated

Robustin committed:
Merge branch 'master' into Future7

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -378,7 +396,210 @@ RCD
+
+/obj/item/weapon/construction/rcd/arcd/afterattack(atom/A, mob/user)
+	range_check(A,user)
+	if((isturf(A) && A.density && mode==RCD_DECONSTRUCT) || (isturf(A) && !A.density) || (istype(A, /obj/machinery/door/airlock) && mode==RCD_DECONSTRUCT) || istype(A, /obj/structure/grille) || (istype(A, /obj/structure/window) && mode==RCD_DECONSTRUCT) || istype(A, /obj/structure/girder))
```
Really put this in a proc if(is_valid_target(A)) ..

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -378,7 +396,210 @@ RCD
+	icon_state = "rld-5"
+	matter = 200
+	max_matter = 200
+	var/mode = 1
```
Defines for modes

Robustin replied:
I knew it, this was your plan all along, baiting me into doing a full RCD refactor like 2 weeks after Goof did a partial one, why didn't anyone tell him to use defines ;_;

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -378,7 +396,210 @@ RCD
+
+/obj/item/weapon/construction/rld/afterattack(atom/A, mob/user)
+	var/turf/start = get_turf(src)
+	if(!(A in view(7, get_turf(start))))
```
You add common range function then not use it.

Robustin replied:
It wasn't common but I'll change it so that it is

Robustin committed:
The update that gets merged but not tested and breaks things

Robustin committed:
Indent memes

ChangelingRain commented:
Sorry dude but literally none of the maintainers want to merge this plus the actual design lead doesn't like it.

KorPhaeron commented:
These are actually cool if he can fix the code

Robustin commented:
I literally got a maintainer to support this because Kor told me its what I needed to keep my PR's open, that was less than 2 days ago, what the fuck.

GunHog commented on tgstation.dme:
```diff
@@ -35,7 +35,6 @@
 #include "code\__DEFINES\hud.dm"
 #include "code\__DEFINES\inventory.dm"
 #include "code\__DEFINES\is_helpers.dm"
-#include "code\__DEFINES\language.dm"
```
You forgot to omit this file from your conflict fix.

Robustin replied:
Its more than that, pulling from upstream removed several files but the rest of the code doesn't reflect that so im getting hundreds of errors.

Robustin committed:
Icon conflicts will be the end of me

Robustin committed:
Merge branch 'master' into Future7

fludd12 commented:
A scary thought - can these be used through camconsoles like bags of holding and such? Or was that fixed ages ago and I'm needlessly complicating things?

Robustin commented:
I specifically tested and addressed that fludd12.

I added an -inview- check that prevents that, but without the range check you could use them through cameras.

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -25,30 +25,115 @@ RCD
+	var/plasteelmultiplier = 3 //Plasteel is worth 3 times more than glass or metal
+	var/no_ammo_message = "<span class='warning'>The \'Low Ammo\' light on the device blinks yellow.</span>"
+
+/obj/item/weapon/construction/New()
```
Change to Initialize

Robustin committed:
Merge branch 'master' into Future7

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -378,7 +403,188 @@ RCD
+/obj/item/weapon/construction/rcd/arcd/afterattack(atom/A, mob/user)
+	range_check(A,user)
+	target_check(A,user)
+	user.Beam(A,icon_state="rped_upgrade",time=30)
```
I think you meant if(target_check(..))user.Beam here.

Robustin replied:
I probably did but I wasn't having issues with the check during testing, odd.

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -378,7 +403,188 @@ RCD
+	..()
+	IconAmmo()
+
+/obj/item/weapon/construction/rld/proc/IconAmmo()
```
Why make this new proc instead of adding to the base update_icon ?

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -378,7 +403,188 @@ RCD
+					user.Beam(A,icon_state="nzcrentrs_power",time=15)
+					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
+					if(do_after(user, decondelay, target = A))
+						if(!useResource(deconcost, user)) return 0
```
Break after if please. It's repository standard.

AnturK commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -378,7 +403,188 @@ RCD
+				G.color = color_choice
+				G.light_color = G.color
+				G.throw_at(A, 9, 3, user)
+				G.on = TRUE
```
Throw_at does not block, if you meant for the light to show only after the throw. You'll have to add callback to throw_at

Robustin replied:
What do you mean does not block? This feature just fires out a glowstick.

AnturK replied:
I mean did you want the glowstick to start glowing when shot or when it hits something ?

Robustin replied:
It really makes no difference since the "throw" is resolved in a fraction of a second anyway.

Robustin committed:
Requested changes

Robustin committed:
GLOB

Robustin committed:
probably my 30th commit b/c of refactors

Robustin committed:
Compiles

Dorsisdwarf commented:
This is by far my favorite engimart PR

Robustin commented:
The RLD is pretty fun, colored lights galore. The RCD is great but a lot more bland, its just a more efficient RCD.

Either way I wish it'd just get merged :\

Robustin committed:
Merge branch 'master' into Future7

Robustin commented:
At this point I think I've spent more time fixing conflicts waiting on this PR than I actually spent coding it. Pretty soon this PR is going to touch every god damn file in the code after enough refactors.

Robustin committed:
Thanks Runtimestation

optimumtact commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -25,30 +25,115 @@ RCD
+
+
+/obj/item/weapon/construction/Destroy()
+	qdel(spark_system)
```
QDEL_NULL(spark_system)

optimumtact replied:
does the nulling automatically.

optimumtact commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -378,7 +403,186 @@ RCD
+
+/obj/item/weapon/construction/rld/ui_action_click(mob/user, var/datum/action/A)
+	if(istype(A, /datum/action/item_action/pick_color))
+		color_choice = input(usr,"Choose Color") as color
```
use user not usr, smh

optimumtact commented on code/game/objects/items/weapons/RCD.dm:
```diff
@@ -378,7 +403,186 @@ RCD
+					playsound(src.loc, 'sound/machines/click.ogg', 50, 1)
+					playsound(src.loc, 'sound/effects/light_flicker.ogg', 50, 1)
+					if(do_after(user, floordelay, target = A))
+						if(!istype(F)) return 0
```
put the return on a new line below hte if, same goes for anywhere else we use this pattern, it's a violation of our style guidelines

optimumtact commented:
This is almost ready, quit fucking around and finish it

Robustin commented:
Fucking around? Despite the insane amount of change I've had to make I've been responsive on them all while waiting up to a week at a time for followup. PLZ

Robustin committed:
REEEEEE
